backend/schema/corpora_schema_h5ad_implementation.md

```diff
@@ -0,0 +1,78 @@
+goal of using this format is to have all metadata stored in the same file as the data, so cellxgene will have access to
+all Corpora metadata within a single file.
+
+The Corpora schema requirements and definitions for the AnnData `X`, `uns`, `obs`, and `obsm` attributes are below.
+
+
+### Note on types
+The types below are python3 types. Note that a python3 `str` is a sequence of Unicode code points, which is stored
+null-terminated and UTF-8-encoded by anndata.
+
+For succinctness, there is a `link` type used below that is a `dict` {"link_name": `str`, "link_url": `str`}.
+
+## `X`
+
+Corpora does not impose any additional constraints on the `X` data matrix. So it may be sparse or dense and any
+numeric `numpy.dtype`.
+
+## `uns`
+
+Recall that `uns` is a mapping with `str`s as keys. Corpora requires the following keys and values in `uns`:
+
+**Key**|**Value Type**|**Notes**
+-----|-----|-----
+title|`str`|Title of the dataset
+contributors|`list` of `dict`s. each `dict` must a "name": `str` item and may optionally have "email": `str` and "institution": `str`|This is the person who gave us the dataset, not necessarily the author or actual data generator.
+layer\_descriptions|dict with string keys and values|One key must be "X" which describes the transformations (if any) performed to produce the X matrix cellxgene displays.
+organism|`str`| 
+organism\_ontology|`str`| 
+project\_name|`str`|Name of the project that contains this dataset
+project\_description|`str`| 
+project\_protocol\_links|`list` of `link`s|May be empty list
+project\_raw\_data\_links|`list` of `link`s|May be empty list
+project\_other\_links|`list` of `link`s|May be empty list
+
+The `project\_protocol\_links`, `project\_raw\_data\_links`, and `project\_other\_links` keys are always present in `uns` as
+they are used in the creation of "projects during Corpora ingestion. But if the values for these keys are empty lists, then
+wrt creating a user interface, that is equivalent to the fields being missing.
+
+Corpora defines the following optional keys and values in `uns`. If they key is present, then the value must not be empty.
+
+**Key**|**Value Type**|**Notes**
+-----|-----|-----
+preprint\_doi|`str`|Suitable for submission to the CrossRef API
+publication\_doi|`str`|Suitable for submission to the CrossRef API
+default\_embedding|`str`|Must match a key to an embedding in `obsm`.
+default\_field|`str`|Must match a column name in `obs`
+tags|`list` of `str`s| 
+<obs\_column>\_colors|`list` of color specifications (see anndata/cellxgene documentation)|<obs\_column> must be a column name in `obs`. There may be multiple keys like this.
+
+
+## `obsm`
+
+`obsm` is a mapping from `str`s to matrices of shape (n, m) where `n` is the number of rows in `X` and `m >= 1`.
+Corpora requires one value in `obms` to be a two-dimensional embedding, meaning it is of shape (n, 2) and is a
```

Comment:
does corpora require `m==2`, or `m>=2`?  

currently cellxgene requires `m>=2`, but ignores anything other than the first two.

Also, for embeddings, I think you want to further restrict the values -- they should be numeric, but also non-Infinite.   NaN and finite values make sense in an embedding - not sure what +/- Inf means, and currently cellxgene will refuse to display any embedding-like column containing these values.

Reply:
In addition, I believe the embedding names must be prefixed with `X_`.

This is covered in the more general Corpora schema doc - you might want to move that information (`Implementations` section) to this doc.

Reply:
I suppose we wouldn't want to say having PCA only breaks the spec, so `m>=2`. I'll add the stuff about `X_` and being finite.